### PR TITLE
[WIP] Refactor work loop a bit

### PR DESCRIPTION
**Pull request by fw42**

### CODE REVIEW

File: lib/resque/worker.rb [hunk 1]
```diff
@@ -204,45 +204,22 @@ def glob_match(pattern)
...
     def work(interval = 5.0, &block)
       interval = Float(interval)
-      $0 = "resque: Starting"
       startup
 
       loop do
         break if shutdown?
 
-        if not paused? and job = reserve
+        if !paused? && job = reserve
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: Keep the `and` since it has a lower precedence than `=`.  For example:

```
irb> a = 1 && b = 2; [a, b]
=> [2, 2]
irb> a = 1 and b = 2; [a, b]
=> [1, 2]
```

File: lib/resque/worker.rb [hunk 2]
```diff
@@ -204,45 +204,22 @@ def glob_match(pattern)
     # has completed processing. Useful for testing.
     def work(interval = 5.0, &block)
       interval = Float(interval)
-      $0 = "resque: Starting"
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: Why did you remove this line?

Comment (fw42) [inline lib/resque/worker.rb]: I moved it into the `startup` method

Comment (fw42) [inline lib/resque/worker.rb]: Oh, looks like I didn't. That's what I meant to do. Good catch.

File: lib/resque/worker.rb (hunk 1 above)
Comment (fw42) [inline lib/resque/worker.rb]: not sure I understand why that matters in this case (the first part is not an assignment)

Comment (dylanahsmith) [inline lib/resque/worker.rb]: The precedence rules around it were confusing (e.g. I was surprised this even worked given the precedence rules) and it won't behave properly if the conditions are reversed or another condition is added.

Comment (fw42) [inline lib/resque/worker.rb]: fair enough

File: lib/resque/worker.rb [hunk 3]
```diff
@@ -253,13 +230,38 @@ def work(interval = 5.0, &block)
...
+      if @child = fork(job)
+        srand # Reseeding
+        procline "Forked #{@child} at #{Time.now.to_i}"
+        begin
+          Process.waitpid(@child)
+        rescue SystemCallError
+          nil
+        end
+        @child = nil
+        job.fail(DirtyExit.new("Child process received unhandled signal #{$?.stopsig}")) if $?.signaled?
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: You haven't set `job.worker = self` on this code path, since you moved that to work_one_job.  Resque::Job#fail does use `worker`.

Comment (fw42) [inline lib/resque/worker.rb]: oh interesting.. good catch. sucks that no test failed because of that.

File: lib/resque/worker.rb [hunk 4]
```diff
@@ -253,13 +230,38 @@ def work(interval = 5.0, &block)
...
+      else
+        unregister_signal_handlers if term_child
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: This will unregister signal handlers on platform that don't support `fork` and don't set `ENV["FORK_PER_JOB"]`.  In that case `will_fork?` will be `true` before the first time `fork` is called, but then `fork` will set `@cant_fork = true` and return `nil` which leads to this code path which unregisters signal handlers and exits.

I guess that doesn't affect us, but it would matter if you tried to get this upstream.

Comment (fw42) [inline lib/resque/worker.rb]: wasn't the old code already doing the same thing? I just removed the `will_fork?` check here because this method is only ever called when `will_fork?` is true anyway.

Comment (dylanahsmith) [inline lib/resque/worker.rb]: Upstream is super confusing, because on platforms that don't support fork a call to `will_fork?` will return `true` before `fork` is called because it only determines whether it can fork by first trying to fork.  However, after that first call to `fork` calls to `will_fork?` will start returning `false`.  Which means that you can't assume that `will_fork?` will return `true` here.

I think I might try to make an upstream pull request to clean that up, since I think this could be cleaned up and avoid a longstanding issue where this code path can be escaped by the child because they aren't using fork with a block.

Comment (fw42) [inline lib/resque/worker.rb]: ❤️ ❤️ ❤️

**Comment by Sirupsen** [comment]
Besides the legendary code review just given by Dylan, I don't have anything to add.
